src/aspire/abinitio/commonline_d2.py

```diff
@@ -0,0 +1,1890 @@
+        Rijs_rows = np.zeros((n_pairs, 3, 3, 3), dtype=self.dtype)
+        for layer in range(3):
+            Rijs_rows[:, layer] = 0.5 * (Rijs[:, 0] + Rijs[:, layer + 1])
+
+        # Partition the set of matrices Rijs_rows into 3 sets of matrices, where
+        # each set there are only matrices Qi^T*Ik*Qj for a unique value of k in
+        # {1,2,3}.
+        # First determine for each pair of tuples of the form {Qi^T*Ik*Qj} and
+        # {Qr^T*Il*Qj} where {i,j}\cap{r,l}==1, whether l==r.
+        color_perms = self._match_colors(Rijs_rows)
+
+        # Compute eigenvectors of color matrix. This is just a matrix of dimensions
+        # 3(N choose 2)x3(N choose 2) where each entry corresponds to a pair of
+        # matrices {Qi^T*Ir*Qj} and {Qr^T*Il*Qj} and equals \delta_rl.
+        # The 2 leading eigenvectors span a linear subspace which contains a
+        # vector which encodes the partition. All the entries of the vector are
+        # either 1,0 or -1, where the number encodes which the index r in Ir.
+        # This vector is a linear combination of the two leading eigen vectors,
+        # and so we 'unmix' these vectors to retrieve it.
+        color_mat = la.LinearOperator(
+            (3 * n_pairs,) * 2, lambda v: self._mult_cmat_by_vec(color_perms, v)
+        )
+        v0 = randn(
+            3 * n_pairs, seed=self.seed
+        )  # Seed eigs initial vector for iterative method
+        v0 = v0 / norm(v0)
+        vals, colors = la.eigs(color_mat, k=3, which="LR", v0=v0)
+        vals = np.real(vals)
+        colors = np.real(colors)
+        colors = np.sign(colors[0]) * colors  # Stable eigs
+        cp, _ = self._unmix_colors(colors[:, :2])
+
+        return cp, Rijs_rows
+
+    def _match_colors(self, Rijs_rows):
+        """
+        Partition the set of matrices Rijs_rows, which correspond to a permutation of
+        the outer products of the m'th rows of Ri and Rj, into 3 sets of matrices each
+        corresponding to an m'th row. Returns the permutations which induce the partition.
+
+        :param Rijs_rows: An n_pairsx3x3x3 array of m'th row outer products for the pairs
+            Ri, Rj, where Rijs_rows[:, i] is the m'th row outer product of unknown row m.
+        :return: n_pairs length array corresponding to the permutation which color matches
+            Rijs_rows.
+        """
+        Rijs_rows_t = np.transpose(Rijs_rows, (0, 1, 3, 2))
+        trip_perms = np.array(
+            [[0, 1, 2], [0, 2, 1], [1, 0, 2], [1, 2, 0], [2, 0, 1], [2, 1, 0]],
+            dtype="int",
+        )
+        inverse_perms = np.array(
+            [
+                [0, 1, 2],
+                [0, 2, 1],
+                [1, 0, 2],
+                [2, 0, 1],
+                [1, 2, 0],
+                [2, 1, 0],
+            ],
+            dtype="int",
+        )
+
+        m = np.zeros((6, 6), dtype=self.dtype)
+        colors_i = np.zeros((len(self.triplets), 3), dtype=self.dtype)  # ints?
```

Comment:
Indeed, why not ints?

Reply:
oops, missed my own comment. Thanks.

Done.